File: 095_SearchInBST.cpp
/*
Problem: Search in BST
Description: Search for a value in a Binary Search Tree.

Approach: Use BST property - go left if smaller, right if larger.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* searchBST(TreeNode* root, int val) {
    if (!root || root->val == val) return root;
    
    if (val < root->val) {
        return searchBST(root->left, val);
    }
    return searchBST(root->right, val);
}

int main() {
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    
    TreeNode* result = searchBST(root, 2);
    if (result) cout << "Found: " << result->val << endl;
    else cout << "Not found" << endl;
    
    return 0;
}

/*
Time Complexity: O(h) where h is height
Space Complexity: O(h) for recursion
*/

File: 095_SearchInBST.py
"""
Problem: Search in BST
Description: Search for a value in a Binary Search Tree.

Approach: Use BST property - go left if smaller, right if larger.
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def searchBST(root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
    if not root or root.val == val:
        return root
    
    if val < root.val:
        return searchBST(root.left, val)
    return searchBST(root.right, val)

def main():
    root = TreeNode(4)
    root.left = TreeNode(2)
    root.right = TreeNode(7)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)
    
    result = searchBST(root, 2)
    if result:
        print(f"Found: {result.val}")
    else:
        print("Not found")

if __name__ == "__main__":
    main()

"""
Time Complexity: O(h) where h is height
Space Complexity: O(h) for recursion
"""
